File: server/app/schemas/delivery.py
from enum import Enum
from uuid import UUID
from datetime import datetime
from pydantic import BaseModel
from typing import Optional
from decimal import Decimal


class DeliveryStatus(str, Enum):
    REQUESTED = "requested"
    ASSIGNED = "assigned"
    EN_ROUTE_PICKUP = "en_route_pickup"
    PICKED_UP = "picked_up"
    EN_ROUTE_DROP = "en_route_drop"
    DELIVERED = "delivered"
    CONFIRMED = "confirmed"


class UpdateDeliveryStatus(BaseModel):
    status: DeliveryStatus


class DeliveryResponse(BaseModel):
    id: UUID
    order_id: UUID
    rider_id: Optional[UUID]
    status: str
    pickup_lat: Optional[float]
    pickup_lng: Optional[float]
    drop_lat: Optional[float]
    drop_lng: Optional[float]
    distance_km: Optional[Decimal]
    fee_pk: int
    created_at: datetime
    
    model_config = {"from_attributes": True}

File: app/lib/data/models/delivery.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'delivery.freezed.dart';
part 'delivery.g.dart';

enum DeliveryStatus {
  @JsonValue('requested')
  requested,
  @JsonValue('assigned')
  assigned,
  @JsonValue('en_route_pickup')
  enRoutePickup,
  @JsonValue('picked_up')
  pickedUp,
  @JsonValue('en_route_drop')
  enRouteDrop,
  @JsonValue('delivered')
  delivered,
  @JsonValue('confirmed')
  confirmed,
}

@freezed
class Delivery with _$Delivery {
  const factory Delivery({
    required String id,
    required String orderId,
    String? riderId,
    required DeliveryStatus status,
    double? pickupLat,
    double? pickupLng,
    double? dropLat,
    double? dropLng,
    double? distanceKm,
    @Default(0) int feePk,
    required DateTime createdAt,
  }) = _Delivery;

  factory Delivery.fromJson(Map<String, dynamic> json) => _$DeliveryFromJson(json);
}

